torch/_refs/__init__.py

```diff
@@ -4278,20 +4273,79 @@ def arange(
 ) -> TensorLikeType:
     utils.check_layout(layout)
     utils.check_pin_memory(pin_memory)
+    device = torch.device(utils.device_or_default(device))
+
+    assert not isinstance(start, complex)
+    assert not isinstance(end, complex)
+    assert not isinstance(step, complex)
+
     # Case: torch.arange(5)
     if end is None:
         end = start
         start = 0
-    return prims.arange(
-        start,
-        end,
-        step,
-        dtype=dtype,
-        # layout=layout,
+    utils.check(step != 0, lambda: "step must be nonzero")
+    utils.check(
+        (step > 0 and end >= start) or (step < 0 and end <= start),
+        lambda: "upper bound and lower bound inconsistent with step sign",
+    )
+
+    def is_finite(x):
+        return not isinstance(x, FloatWithoutSymFloat) or math.isfinite(x)
+
+    utils.check(
+        is_finite(start) and is_finite(end),
+        lambda: f"unsupported range: {start} -> {end}",
+    )
+    utils.check(
+        is_finite(step),
+        lambda: f"step must be finite but got {step}",
+    )
+
+    if dtype is None:
+        args = (start, end, step)
+        integer_args = builtins.all(isinstance(arg, IntLike) for arg in args)
+        dtype = torch.int64 if integer_args else torch.get_default_dtype()
+
+    is_integer = utils.is_integer_dtype(dtype)
+    if is_integer:
+        xstart = sym_int(start)
+        xend = sym_int(end)
+        xstep = sym_int(step)
+
+    # For int64 we truncate arguments to int before calculating length, but
+    # other integral dtypes we don't. Weird... but needed to match ATen shapes.
+    if dtype == torch.int64:
+        length = math.ceil((xend - xstart) / xstep)
+    else:
+        print(start, end, step)
+        length = math.ceil((end - start) / step)
+
+    if is_integer:
+        return prims.iota(
+            length,
+            start=xstart,
+            step=xstep,
+            dtype=dtype,
+            device=device,
+            requires_grad=requires_grad,
+        )
+
+    computation_dtype = utils.get_acc_type(dtype, device)
```

Comment:
Is this not the accumulation type used by reductions?

Reply:
Yes and t's also used by several other operators including `arange`:
https://github.com/pytorch/pytorch/blob/3e6978172e86a97dc4c248406d2b2509a837299f/aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp#L23

Reply:
Can't this then be done via `utils.reduction_dtypes`?

Reply:
That doesn't seem to promote to double on cpu, or am I missing something?

Reply:
ag, I see, it does not. I thought `acc_type` had been deprecated in favour of `opmath_type`...